### PR TITLE
[bitnami/kafka] Add separate service account for kafka exporter

### PR DESCRIPTION
**Pull request by nflaig**
**Description of the change**

This PR adds a separate service account and rbac for Kafka exporter.

**Benefits**

Right now there is only one service account for both the Kafka pods and the Kafka Exporter pods. This leads to limitations if the service accounts need to be different. An example case would be if you are using Consul operator, the service account name needs to match the service name in Consul. If you want to register Kafka and Kafka Exporter in Consul you need two different service accounts.

Also the kafka exporter should be considered a separate service and therefore the service accounts should be separated.

**Possible drawbacks**

There are no changes for users that use the default settings but there might be changes required for users that explicitly disabled the service account before.

**Additional information**

Implements the changes discussed in #7759.

**Checklist** 

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. [bitnami/<name_of_the_chart>])

### CODE REVIEW

File: bitnami/kafka/values.yaml [hunk 1]
```diff
@@ -942,6 +942,21 @@ metrics:
       ##   - myRegistryKeySecretName
       ##
       pullSecrets: []
+    ## Kafka exporter pods ServiceAccount
+    ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/
+    ##
+    serviceAccount:
+      ## @param metrics.kafka.serviceAccount.create Enable creation of ServiceAccount for Kafka exporter pods
+      ##
+      create: true
```
Comment (marcosbc) [inline bitnami/kafka/values.yaml:951]: I would recommend to change the default to `false`, as this could cause problems. Any concerns against that?

**Review by marcosbc — CHANGES_REQUESTED** [review]
Thanks for the PR! Please have a look at my comment

File: bitnami/kafka/templates/kafka-metrics-role.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{{- if .Values.rbac.create -}}
```
Comment (juan131) [inline bitnami/kafka/templates/kafka-metrics-role.yaml:1]: This condition is incomplete
```suggestion
{{- if and .Values.rbac.create .Values.metrics.kafka.enabled Values.metrics.kafka.serviceAccount.create  }}
```

File: bitnami/kafka/templates/kafka-metrics-rolebinding.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{{- if and .Values.metrics.kafka.serviceAccount.create .Values.rbac.create }}
```
Comment (juan131) [inline bitnami/kafka/templates/kafka-metrics-rolebinding.yaml:1]: Same here:
```suggestion
{{- if and .Values.rbac.create .Values.metrics.kafka.enabled Values.metrics.kafka.serviceAccount.create  }}
```

File: bitnami/kafka/templates/kafka-metrics-serviceaccount.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{{- if .Values.metrics.kafka.serviceAccount.create }}
```
Comment (juan131) [inline bitnami/kafka/templates/kafka-metrics-serviceaccount.yaml:1]: There's no sense on creating the SA if we're not enabling Kafka metrics:
```suggestion
{{- if and .Values.metrics.kafka.enabled Values.metrics.kafka.serviceAccount.create  }}
```

File: bitnami/kafka/values.yaml (hunk 1 above)
Comment (juan131) [reply bitnami/kafka/values.yaml:951]: I agree. As far as RBAC isn't enabled, it shouldn't make much a difference

File: bitnami/kafka/templates/kafka-metrics-role.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  {{- end }}
+rules:
+  - apiGroups:
+      - ""
+    resources:
+      - services
+    verbs:
+      - get
+      - list
+      - watch
```
Comment (juan131) [inline bitnami/kafka/templates/kafka-metrics-role.yaml:22]: Could you please justify why these permissions are required?

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks for the PR, please check my comments

File: bitnami/kafka/templates/kafka-metrics-role.yaml (hunk 5 above)
Comment (nflaig) [reply bitnami/kafka/templates/kafka-metrics-role.yaml:22]: Those are the same permissions as previously applied to the kafka exporter by using the same role/service account as kafka. I am not exactly sure what permissions are required here but I am happy to reduce it further if you have better insights on what is actually required

File: bitnami/kafka/templates/kafka-metrics-role.yaml (hunk 2 above)
Comment (nflaig) [reply bitnami/kafka/templates/kafka-metrics-role.yaml]: Good catch I totally missed that

File: bitnami/kafka/templates/kafka-metrics-rolebinding.yaml (hunk 3 above)
Comment (nflaig) [reply bitnami/kafka/templates/kafka-metrics-rolebinding.yaml]: fixed

File: bitnami/kafka/templates/kafka-metrics-serviceaccount.yaml (hunk 4 above)
Comment (nflaig) [reply bitnami/kafka/templates/kafka-metrics-serviceaccount.yaml]: fixed

**Comment by nflaig** [comment]
Thanks guys for the review! I fixed the issues mentioned in the comments

File: bitnami/kafka/templates/kafka-metrics-role.yaml (hunk 5 above)
Comment (juan131) [reply bitnami/kafka/templates/kafka-metrics-role.yaml:22]: These permissions were set for a very specific purpose: loadbalancer/nodeports auto-discovery when enabling [external access](https://github.com/bitnami/charts/tree/master/bitnami/kafka#accessing-kafka-brokers-from-outside-the-cluster). In the `auto-discovery.sh` script you can find some commands that use `kubectl` to access the K8s API and retrieve information about existing services:

- https://github.com/bitnami/charts/blob/master/bitnami/kafka/templates/scripts-configmap.yaml#L45

However, I don't see any reason to add these permissions to the Kafka metrics pods.

Comment (nflaig) [reply bitnami/kafka/templates/kafka-metrics-role.yaml:22]: Yes, it does not make sense to add the to the kafka exporter pods. The question then is do we need a role at all? Mostly the requirement is that the kafka exporter pods have separate service account and not use the same as the kafka pods

Comment (juan131) [reply bitnami/kafka/templates/kafka-metrics-role.yaml:22]: > do we need a role at all?

I don't think so. I think it's safe to remove the Role/RoleBinding for this use case.

**Review by juan131 — APPROVED** [review]
LGTM! Thanks for implementing the changes

File: bitnami/kafka/templates/_helpers.tpl [hunk 6]
```diff
@@ -60,6 +60,24 @@ Return the proper image name (for the init container volume-permissions image)
...
+{{- define "kafka.metrics.kafka.name" -}}
+  {{- printf "%s-exporter" (include "kafka.fullname" .) | trunc 63 | trimSuffix "-" }}
+{{- end -}}
+
+{{/*
+ Create the name of the service account to use for Kafka exporer pods
+ */}}
+{{- define "kafka.metrics.kafka.serviceAccountName" -}}
+{{- if .Values.metrics.kafka.serviceAccount.create -}}
+    {{ default (include "kafka.metrics.kafka.name" .) .Values.metrics.kafka.serviceAccount.name }}
```
Comment (juan131) [inline bitnami/kafka/templates/_helpers.tpl:75]:
```suggestion
    {{ default (include "kafka.metrics.kafka.fullname" .) .Values.metrics.kafka.serviceAccount.name }}
```

File: bitnami/kafka/templates/kafka-metrics-deployment.yaml [hunk 7]
```diff
@@ -7,7 +7,7 @@
 apiVersion: {{ include "common.capabilities.deployment.apiVersion" . }}
 kind: Deployment
 metadata:
-  name: {{ template "kafka.fullname" . }}-exporter
+  name: {{ template "kafka.metrics.kafka.name" . }}
```
Comment (juan131) [inline bitnami/kafka/templates/kafka-metrics-deployment.yaml:10]:
```suggestion
  name: {{ template "kafka.metrics.kafka.fullname" . }}
```

**Review by juan131 — COMMENTED** [review]
Super minor change

File: bitnami/kafka/templates/_helpers.tpl [hunk 8]
```diff
@@ -60,6 +60,25 @@ Return the proper image name (for the init container volume-permissions image)
 {{ include "common.images.image" (dict "imageRoot" .Values.volumePermissions.image "global" .Values.global) }}
 {{- end -}}
 
+{{/*
+Create a default fully qualified Kafka exporter name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+*/}
```
Comment (juan131) [inline bitnami/kafka/templates/_helpers.tpl:66]:
```suggestion
*/}}
```

**Review by juan131 — COMMENTED** [review]
Typo

**Pull request merged**